Respond to thermal pressure more aggressively

bnham opened the pull request:
#### 984773adbeed0269b4c5bdf1d1938afdffef343a
<pre>
Respond to thermal pressure more aggressively
<a href="https://bugs.webkit.org/show_bug.cgi?id=284586">https://bugs.webkit.org/show_bug.cgi?id=284586</a>
<a href="https://rdar.apple.com/141402176">rdar://141402176</a>

Reviewed by Chris Dumez.

Currently WebKit responds to thermal pressure only by increasing the DOMTimer alignment interval.

This patch adds a preference (RespondToThermalPressureAggressively) which causes to WebKit to
respond to thermal pressure by taking additional actions:

 - Reducing the preferred animation frame rate
 - Disabling autoplay videos without a user gesture
 - Disabling HDR
 - Disabling speculative loads in NetworkProcess

These actions should be safe since we throttle these exact behaviors when enabling low power mode.

* LayoutTests/fast/animation/css-animation-throttling-expected.txt: Renamed from LayoutTests/fast/animation/css-animation-throttling-lowPowerMode-expected.txt.
* LayoutTests/fast/animation/css-animation-throttling.html: Renamed from LayoutTests/fast/animation/css-animation-throttling-lowPowerMode.html.
* LayoutTests/fast/animation/request-animation-frame-throttling-aggressiveThermalMitigation-expected.txt: Added.
* LayoutTests/fast/animation/request-animation-frame-throttling-aggressiveThermalMitigation.html: Added.
* LayoutTests/fast/dom/timer-throttling-aggressiveThermalMitigation-expected.txt: Added.
* LayoutTests/fast/dom/timer-throttling-aggressiveThermalMitigation.html: Added.
* LayoutTests/media/media-usage-state-expected.txt:
* LayoutTests/media/media-usage-state.html:
* LayoutTests/media/modern-media-controls/start-support/start-support-aggressiveThermalMitigation-expected.txt: Added.
* LayoutTests/media/modern-media-controls/start-support/start-support-aggressiveThermalMitigation.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/svg/animations/animations-throttling-expected.txt: Renamed from LayoutTests/svg/animations/animations-lowPowerMode-throttling-expected.txt.
* LayoutTests/svg/animations/animations-throttling.html: Renamed from LayoutTests/svg/animations/animations-lowPowerMode-throttling.html.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::initializeMediaSession):
(WebCore::HTMLMediaElement::shouldForceControlsDisplay const):
(WebCore::HTMLMediaElement::removeBehaviorRestrictionsAfterFirstUserGesture):
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::restrictionNames):
(WebCore::MediaElementSession::playbackStateChangePermitted const):
(WebCore::MediaElementSession::updateMediaUsageIfChanged):
* Source/WebCore/html/MediaElementSession.h:
* Source/WebCore/page/Page.cpp:
(WebCore::m_activeNowPlayingSessionUpdateTimer):
(WebCore::Page::setAggressiveThermalMitigationEnabledForTesting):
(WebCore::Page::handleThermalMitigationChange):
* Source/WebCore/page/Page.h:
(WebCore::Page::isAggressiveThermalMitigationEnabled const):
* Source/WebCore/platform/cocoa/PowerSourceNotifier.h:
* Source/WebCore/platform/graphics/AnimationFrameRate.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/AnimationFrameRate.h:
* Source/WebCore/platform/graphics/MediaUsageInfo.h:
* Source/WebCore/platform/mac/PlatformScreenMac.mm:
(WebCore::collectScreenProperties):
* Source/WebCore/svg/animation/SMILTimeContainer.cpp:
(WebCore::SMILTimeContainer::animationFrameDelay const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::resetToConsistentState):
(WebCore::Internals::setAggressiveThermalMitigationEnabled):
(WebCore::Internals::setMediaElementRestrictions):
(WebCore::Internals::mediaUsageState const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::Cache::Cache):
(WebKit::NetworkCache::Cache::shouldUseSpeculativeLoadManager const):
(WebKit::NetworkCache::Cache::updateSpeculativeLoadManagerEnabledState):
* Source/WebKit/NetworkProcess/cache/NetworkCache.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Tools/TestWebKitAPI/Tests/WebCore/AnimationFrameRate.cpp:
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFrameIntervalWithUnspecifiedNominalFramesPerSecond)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFrameIntervalWithFullSpeedNominalFramesPerSecond)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFrameIntervalWith144FPSNominalFramesPerSecond)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFrameIntervalWith120FPSNominalFramesPerSecond)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFrameIntervalWith90FPSNominalFramesPerSecond)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFrameIntervalWith48FPSNominalFramesPerSecond)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFrameIntervalWith30FPSNominalFramesPerSecond)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFramesPerSecondMatchNominalFrameRate)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFramesPerSecondTarget60FPS)):

Canonical link: <a href="https://commits.webkit.org/287858@main">https://commits.webkit.org/287858@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/61f0a8a5a02181df14d35006109fe76f3053b460

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81037 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/562 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34980 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85566 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32023 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83147 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/580 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8364 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63258 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21021 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84106 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/342 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73756 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43556 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/240 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27917 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30481 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/74016 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71786 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28469 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87001 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/80095 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8267 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5834 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71563 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8444 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69591 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70799 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17643 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14840 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13765 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/102501 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8228 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13751 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24900 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8065 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11585 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9873 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
```diff
@@ -111,20 +112,15 @@ Cache::Cache(NetworkProcess& networkProcess, const String& storageDirectory, Ref
+                updateSpeculativeLoadManagerEnabledState();
         });
-        if (!m_lowPowerModeNotifier->isLowPowerModeEnabled())
+        m_thermalMitigationNotifier = makeUnique<WebCore::ThermalMitigationNotifier>([this, weakThis = WeakPtr { *this }](bool) {
```
Do we have release logging when we apply (aggressive) thermal mitigations?

bnham replied:
Nope. I'll add some logging.

bnham committed:
Respond to thermal pressure more aggressively
https://bugs.webkit.org/show_bug.cgi?id=284586
rdar://141402176

Reviewed by Chris Dumez.

Currently WebKit responds to thermal pressure only by increasing the DOMTimer alignment interval.

This patch adds a preference (RespondToThermalPressureAggressively) which causes to WebKit to
respond to thermal pressure by taking additional actions:

 - Reducing the preferred animation frame rate
 - Disabling autoplay videos without a user gesture
 - Disabling HDR
 - Disabling speculative loads in NetworkProcess

These actions should be safe since we throttle these exact behaviors when enabling low power mode.

* LayoutTests/fast/animation/css-animation-throttling-expected.txt: Renamed from LayoutTests/fast/animation/css-animation-throttling-lowPowerMode-expected.txt.
* LayoutTests/fast/animation/css-animation-throttling.html: Renamed from LayoutTests/fast/animation/css-animation-throttling-lowPowerMode.html.
* LayoutTests/fast/animation/request-animation-frame-throttling-aggressiveThermalMitigation-expected.txt: Added.
* LayoutTests/fast/animation/request-animation-frame-throttling-aggressiveThermalMitigation.html: Added.
* LayoutTests/fast/dom/timer-throttling-aggressiveThermalMitigation-expected.txt: Added.
* LayoutTests/fast/dom/timer-throttling-aggressiveThermalMitigation.html: Added.
* LayoutTests/media/media-usage-state-expected.txt:
* LayoutTests/media/media-usage-state.html:
* LayoutTests/media/modern-media-controls/start-support/start-support-aggressiveThermalMitigation-expected.txt: Added.
* LayoutTests/media/modern-media-controls/start-support/start-support-aggressiveThermalMitigation.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/svg/animations/animations-throttling-expected.txt: Renamed from LayoutTests/svg/animations/animations-lowPowerMode-throttling-expected.txt.
* LayoutTests/svg/animations/animations-throttling.html: Renamed from LayoutTests/svg/animations/animations-lowPowerMode-throttling.html.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::initializeMediaSession):
(WebCore::HTMLMediaElement::shouldForceControlsDisplay const):
(WebCore::HTMLMediaElement::removeBehaviorRestrictionsAfterFirstUserGesture):
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::restrictionNames):
(WebCore::MediaElementSession::playbackStateChangePermitted const):
(WebCore::MediaElementSession::updateMediaUsageIfChanged):
* Source/WebCore/html/MediaElementSession.h:
* Source/WebCore/page/Page.cpp:
(WebCore::m_activeNowPlayingSessionUpdateTimer):
(WebCore::Page::setAggressiveThermalMitigationEnabledForTesting):
(WebCore::Page::handleThermalMitigationChange):
* Source/WebCore/page/Page.h:
(WebCore::Page::isAggressiveThermalMitigationEnabled const):
* Source/WebCore/platform/cocoa/PowerSourceNotifier.h:
* Source/WebCore/platform/graphics/AnimationFrameRate.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/graphics/AnimationFrameRate.h:
* Source/WebCore/platform/graphics/MediaUsageInfo.h:
* Source/WebCore/platform/mac/PlatformScreenMac.mm:
(WebCore::collectScreenProperties):
* Source/WebCore/svg/animation/SMILTimeContainer.cpp:
(WebCore::SMILTimeContainer::animationFrameDelay const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::resetToConsistentState):
(WebCore::Internals::setAggressiveThermalMitigationEnabled):
(WebCore::Internals::setMediaElementRestrictions):
(WebCore::Internals::mediaUsageState const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::Cache::Cache):
(WebKit::NetworkCache::Cache::shouldUseSpeculativeLoadManager const):
(WebKit::NetworkCache::Cache::updateSpeculativeLoadManagerEnabledState):
* Source/WebKit/NetworkProcess/cache/NetworkCache.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Tools/TestWebKitAPI/Tests/WebCore/AnimationFrameRate.cpp:
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFrameIntervalWithUnspecifiedNominalFramesPerSecond)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFrameIntervalWithFullSpeedNominalFramesPerSecond)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFrameIntervalWith144FPSNominalFramesPerSecond)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFrameIntervalWith120FPSNominalFramesPerSecond)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFrameIntervalWith90FPSNominalFramesPerSecond)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFrameIntervalWith48FPSNominalFramesPerSecond)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFrameIntervalWith30FPSNominalFramesPerSecond)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFramesPerSecondMatchNominalFrameRate)):
(TestWebKitAPI::TEST(AnimationFrameRate, preferredFramesPerSecondTarget60FPS)):

Canonical link: https://commits.webkit.org/287858@main
